Kernel/FileSystem/Ext2FS/Inode.cpp

```diff
@@ -36,419 +37,317 @@ static u8 to_ext2_file_type(mode_t mode)
+        doubly_indirect_block_contents[offset_in_doubly_indirect_block] = TRY(allocate_and_zero_block());
+        TRY(fs().write_block(m_raw_inode.i_block[EXT2_DIND_BLOCK], doubly_indirect_block_buffer, block_size));
     }
 
-    // Write out the triply indirect block.
-    return fs().write_block(block, buffer, block_contents.size());
-}
+    TRY(fs().read_block(doubly_indirect_block_contents[offset_in_doubly_indirect_block], &singly_indirect_block_buffer, block_size, 0));
 
-ErrorOr<void> Ext2FSInode::shrink_triply_indirect_block(BlockBasedFileSystem::BlockIndex block, size_t old_blocks_length, size_t new_blocks_length, unsigned& meta_blocks)
-{
-    auto const entries_per_block = EXT2_ADDR_PER_BLOCK(&fs().super_block());
-    auto const entries_per_doubly_indirect_block = entries_per_block * entries_per_block;
-    auto const entries_per_triply_indirect_block = entries_per_doubly_indirect_block * entries_per_block;
-    auto const old_triply_indirect_blocks_length = ceil_div(old_blocks_length, entries_per_doubly_indirect_block);
-    auto const new_triply_indirect_blocks_length = new_blocks_length / entries_per_doubly_indirect_block;
-    VERIFY(old_blocks_length > 0);
-    VERIFY(old_blocks_length >= new_blocks_length);
-    VERIFY(new_blocks_length <= entries_per_triply_indirect_block);
-
-    auto block_contents = TRY(ByteBuffer::create_uninitialized(fs().logical_block_size()));
-    auto* block_as_pointers = (unsigned*)block_contents.data();
-    auto buffer = UserOrKernelBuffer::for_kernel_buffer(reinterpret_cast<u8*>(block_as_pointers));
-    TRY(fs().read_block(block, &buffer, fs().logical_block_size()));
-
-    // Shrink the doubly indirect blocks.
-    for (unsigned i = new_triply_indirect_blocks_length; i < old_triply_indirect_blocks_length; i++) {
-        auto const processed_blocks = i * entries_per_doubly_indirect_block;
-        auto const old_doubly_indirect_blocks_length = min(old_blocks_length > processed_blocks ? old_blocks_length - processed_blocks : 0, entries_per_doubly_indirect_block);
-        auto const new_doubly_indirect_blocks_length = min(new_blocks_length > processed_blocks ? new_blocks_length - processed_blocks : 0, entries_per_doubly_indirect_block);
-        dbgln_if(EXT2_BLOCKLIST_DEBUG, "Ext2FSInode[{}]::shrink_triply_indirect_block(): Shrinking doubly indirect block {} at index {}", identifier(), block_as_pointers[i], i);
-        TRY(shrink_doubly_indirect_block(block_as_pointers[i], old_doubly_indirect_blocks_length, new_doubly_indirect_blocks_length, meta_blocks));
+    singly_indirect_block_contents[offset_in_singly_indirect_block] = on_disk_index.value();
+    TRY(fs().write_block(doubly_indirect_block_contents[offset_in_doubly_indirect_block], singly_indirect_block_buffer, block_size));
+
+    if (on_disk_index != 0)
+        return {};
+
+    for (size_t i = 0; i < entries_per_block; ++i) {
+        if (singly_indirect_block_contents[i] != 0)
+            return {};
```

Comment:
With a `Span` you could use either `contains_slow` or `AK::any_of` to make this a bit more concise